src/providers/token-fee-provider.ts

```diff
@@ -0,0 +1,164 @@
+import { ChainId, Token } from '@uniswap/sdk-core';
+import { BigNumber } from '@ethersproject/bignumber';
+import _ from 'lodash';
+
+import { TokenFeeDetector__factory } from '../types/other/factories/TokenFeeDetector__factory';
+import { log, WRAPPED_NATIVE_CURRENCY } from '../util';
+
+import { ICache } from './cache';
+import { IMulticallProvider } from './multicall-provider';
+import { ProviderConfig } from './provider';
+import { DEFAULT_ALLOWLIST } from './token-validator-provider';
+
+export type TokenFeeResult = {
+  buyFeeBps: BigNumber,
+  sellFeeBps: BigNumber,
+}
+
+// on detector failure, assume no fee
+const DEFAULT_TOKEN_FEE_RESULT = {
+  buyFeeBps: BigNumber.from(0),
+  sellFeeBps: BigNumber.from(0),
+};
+
+export interface TokenFeeResults {
+  getFeesByToken(token: Token): TokenFeeResult | undefined;
+}
+
+// address at which the FeeDetector lens is deployed
+const FEE_DETECTOR_ADDRESS = '0x57eC54d113719dDE9A90E6bE807524a86560E89D';
+// Amount has to be big enough to avoid rounding errors, but small enough that
+// most v2 pools will have at least this many token units
+// 10000 is the smallest number that avoids rounding errors in bps terms
+const AMOUNT_TO_FLASH_BORROW = '10000';
+// 1M gas limit per validate call, should cover most swap cases
+const GAS_LIMIT_PER_VALIDATE = 1_000_000;
+
+/**
+ * Provider for getting token fee data.
+ *
+ * @export
+ * @interface ITokenFeeProvider
+ */
+export interface ITokenFeeProvider {
+  /**
+   * Gets the fees for the token at each address.
+   *
+   * @param tokens The token addresses to fetch fees for.
+   * @param [providerConfig] The provider config.
+   * @returns A token accessor with methods for accessing the tokens.
+   */
+  validateTokens(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenFeeResults>;
+}
+
+export class TokenFeeProvider implements ITokenFeeProvider {
+  private CACHE_KEY = (chainId: ChainId, address: string) =>
+    `token-${chainId}-${address}`;
+
+  private BASE_TOKEN: string;
+
+  constructor(
+    private chainId: ChainId,
+    private multicall2Provider: IMulticallProvider,
+    private tokenFeeCache: ICache<TokenFeeResult>,
```

Comment:
As I'm working on the router backend support for FOT, I realized that this `TokenFeeProvider` is tightly coupled with the in-memory `ICache`. We will need both in-memory memory (for client-side fallback), as well as dynamo caching (for routing-api distributed caches).

I think I will create another PR, meanwhile leveraging the insights and feedbacks from this PR.